### PR TITLE
[ATL] Initial implementation of CAtlExeModuleT.

### PR DESCRIPTION
**Pull request by learn-more**
CORE-13908

JIRA issue: [CORE-13908](https://jira.reactos.org/browse/CORE-13908)

### CODE REVIEW

File: sdk/lib/atl/atlbase.h [hunk 1]
```diff
@@ -190,15 +190,18 @@ struct _ATL_REGMAP_ENTRY
     LPCOLESTR szData;
 };
 
-HRESULT __stdcall AtlWinModuleInit(_ATL_WIN_MODULE *pWinModule);
```
Comment (learn-more) [inline sdk/lib/atl/atlbase.h]: sorry couldnt help myself :)

File: sdk/lib/atl/atlbase.h [hunk 2]
```diff
@@ -662,6 +665,127 @@ class CAtlDllModuleT : public CAtlModuleT<T>
...
 
+
+template <class T>
+class CAtlExeModuleT : public CAtlModuleT<T>
+{
+public:
+    DWORD m_dwMainThreadID;
+    //DWORD m_dwTimeOut;
+    //DWORD m_dwPause;
+    //bool m_bDelayShutdown;
```
Comment (learn-more) [inline sdk/lib/atl/atlbase.h]: These 4 are documented, just didnt need them yet.

File: sdk/lib/atl/atlbase.h [hunk 3]
```diff
@@ -662,6 +665,127 @@ class CAtlDllModuleT : public CAtlModuleT<T>
...
+        if (lRet == 0)
+        {
+            ::PostThreadMessage(m_dwMainThreadID, WM_QUIT, 0, 0);
+        }
+        return lRet;
+    }
+
+    bool ParseCommandLine(LPCTSTR lpCmdLine, HRESULT* pnRetCode)
+    {
+        // unimplemented!
```
Comment (learn-more) [inline sdk/lib/atl/atlbase.h]: I did not need it yet.

**Comment by sanchaez** [comment]
Plz test building it with GCC next time ;P

File: sdk/lib/atl/atlbase.h [hunk 4]
```diff
@@ -662,6 +665,127 @@ class CAtlDllModuleT : public CAtlModuleT<T>
     }
 };
 
+
+template <class T>
+class CAtlExeModuleT : public CAtlModuleT<T>
```
Comment (HBelusca) [inline sdk/lib/atl/atlbase.h]: Public declaration has an attribute ATL_NO_VTABLE (which is declared somewhere in our code). I don't know whether it is meaningful to add it there too.

File: sdk/lib/atl/atlbase.h [hunk 5]
```diff
@@ -662,6 +665,127 @@ class CAtlDllModuleT : public CAtlModuleT<T>
...
+{
+public:
+    DWORD m_dwMainThreadID;
+    //DWORD m_dwTimeOut;
+    //DWORD m_dwPause;
+    //bool m_bDelayShutdown;
+
+    CAtlExeModuleT()
```
Comment (HBelusca) [inline sdk/lib/atl/atlbase.h]: Do we usually put the exception "qualifiers" (like throw() ) at the end of member functions? If so we may say that both the ctor and dtor have throw()

File: sdk/lib/atl/atlbase.h (hunk 3 above)
Comment (HBelusca) [reply sdk/lib/atl/atlbase.h]: MSDN says:
"This method is called from CAtlExeModuleT::WinMain and can be overridden to handle command-line switches. The default implementation checks for /RegServer and /UnRegServer command-line arguments and performs registration or unregistration."
https://docs.microsoft.com/en-us/cpp/atl/reference/catlexemodulet-class#parsecommandline

File: sdk/lib/atl/atlbase.h [hunk 6]
```diff
@@ -662,6 +665,127 @@ class CAtlDllModuleT : public CAtlModuleT<T>
...
+
+
+    HRESULT Run(int nShowCmd = SW_HIDE)
+    {
+        HRESULT hr = S_OK;
+
+        T* pThis = static_cast<T*>(this);
+        hr = pThis->PreMessageLoop(nShowCmd);
+
+        if (hr == S_OK)
```
Comment (HBelusca) [inline sdk/lib/atl/atlbase.h]: Depending on your preferences, to decrease the indentation level, one can also do:
```c++
if (hr != S_OK)
    return hr;

pThis->RunMessageLoop();
return pThis->PostMessageLoop();
```

File: sdk/lib/atl/atlbase.h [hunk 7]
```diff
@@ -1685,6 +1809,55 @@ inline HRESULT WINAPI AtlComModuleUnregisterServer(_ATL_COM_MODULE *mod, BOOL bU
...
+    _ATL_OBJMAP_ENTRY **iter;
+    IUnknown* unk = NULL;
+    HRESULT hr;
+
+    if (!module)
+        return E_INVALIDARG;
+
+    for (iter = module->m_ppAutoObjMapFirst; iter < module->m_ppAutoObjMapLast; iter++)
+    {
+        if(!(*iter)->pfnGetClassObject)
```
Comment (HBelusca) [inline sdk/lib/atl/atlbase.h]: Space after "if".

File: sdk/lib/atl/atlbase.h (hunk 3 above)
Comment (learn-more) [reply sdk/lib/atl/atlbase.h]: And your point is?

File: sdk/lib/atl/atlbase.h (hunk 4 above)
Comment (learn-more) [reply sdk/lib/atl/atlbase.h]: We have that on no class yet, but it should be on a ton.
Didn't want to break style

File: sdk/lib/atl/atlbase.h (hunk 5 above)
Comment (learn-more) [reply sdk/lib/atl/atlbase.h]: ditto

File: sdk/lib/atl/atlbase.h (hunk 6 above)
Comment (learn-more) [reply sdk/lib/atl/atlbase.h]: Indentation level is not 300 deep here, so it's quite readable as-is.

File: sdk/lib/atl/atlbase.h [hunk 8]
```diff
@@ -662,6 +665,127 @@ class CAtlDllModuleT : public CAtlModuleT<T>
...
+{
+public:
+    DWORD m_dwMainThreadID;
+    //DWORD m_dwTimeOut;
+    //DWORD m_dwPause;
+    //bool m_bDelayShutdown;
+
+    CAtlExeModuleT()
+    {
+        m_dwMainThreadID = ::GetCurrentThreadId();
```
Comment (ThFabba) [inline sdk/lib/atl/atlbase.h]: A member initializer list should be preferred over assignments.

File: sdk/lib/atl/atlbase.h [hunk 9]
```diff
@@ -662,6 +665,127 @@ class CAtlDllModuleT : public CAtlModuleT<T>
...
+
+    ~CAtlExeModuleT()
+    {
+    }
+
+    int WinMain(int nShowCmd)
+    {
+        HRESULT hr = T::InitializeCom();
+        if (FAILED(hr))
+            return hr;
```
Comment (ThFabba) [inline sdk/lib/atl/atlbase.h]: Shouldn't this function return 0/1 or something?

File: sdk/lib/atl/atlbase.h [hunk 10]
```diff
@@ -662,6 +665,127 @@ class CAtlDllModuleT : public CAtlModuleT<T>
...
+    HRESULT PreMessageLoop(int nShowCmd)
+    {
+        T* pThis = static_cast<T*>(this);
+        return pThis->RegisterClassObjects(CLSCTX_LOCAL_SERVER, REGCLS_MULTIPLEUSE);
+    }
+
+    void RunMessageLoop()
+    {
+        MSG msg;
+        while (GetMessage(&msg, 0, 0, 0))
```
Comment (ThFabba) [inline sdk/lib/atl/atlbase.h]: GetMessage can return -1.

File: sdk/lib/atl/atlbase.h [hunk 11]
```diff
@@ -1685,6 +1809,55 @@ inline HRESULT WINAPI AtlComModuleUnregisterServer(_ATL_COM_MODULE *mod, BOOL bU
...
+            continue;
+
+        hr = (*iter)->pfnGetClassObject((void*)(*iter)->pfnCreateInstance, IID_IUnknown, (void**)&unk);
+        if (FAILED(hr))
+            return hr;
+
+        hr = CoRegisterClassObject(*(*iter)->pclsid, unk, context, flags, &(*iter)->dwRegister);
+        unk->Release();
+        if (FAILED(hr))
+            return hr;
```
Comment (ThFabba) [inline sdk/lib/atl/atlbase.h]: Should this function rollback registration of previous objects if it fails in the middle of the loop?

File: sdk/lib/atl/atlbase.h [hunk 12]
```diff
@@ -1685,6 +1809,55 @@ inline HRESULT WINAPI AtlComModuleUnregisterServer(_ATL_COM_MODULE *mod, BOOL bU
...
+    HRESULT hr;
+
+    if (!module)
+        return E_INVALIDARG;
+
+    for (iter = module->m_ppAutoObjMapFirst; iter < module->m_ppAutoObjMapLast; iter++)
+    {
+        hr = CoRevokeClassObject((*iter)->dwRegister);
+        if (FAILED(hr))
+            return hr;
```
Comment (ThFabba) [inline sdk/lib/atl/atlbase.h]: Similarly to the above, should this continue even on failure?

File: sdk/lib/atl/atlbase.h (hunk 8 above)
Comment (learn-more) [reply sdk/lib/atl/atlbase.h]: Yeah in this case it makes sense.

File: sdk/lib/atl/atlbase.h (hunk 9 above)
Comment (learn-more) [reply sdk/lib/atl/atlbase.h]: Usually 0 means success, so S_OK maps perfectly here.

File: sdk/lib/atl/atlbase.h (hunk 10 above)
Comment (learn-more) [reply sdk/lib/atl/atlbase.h]: Right, will fix.

File: sdk/lib/atl/atlbase.h (hunk 7 above)
Comment (learn-more) [reply sdk/lib/atl/atlbase.h]: Yeah converted the rest from the style of dll/win32/atl/atl.c, missed this one.

File: sdk/lib/atl/atlbase.h (hunk 11 above)
Comment (learn-more) [reply sdk/lib/atl/atlbase.h]: I stole the code from dll/win32/atl/atl.c, did not verify the logic.

Comment (learn-more) [reply sdk/lib/atl/atlbase.h]: After further consideration I think it's fine as-is.

File: sdk/lib/atl/atlbase.h (hunk 9 above)
Comment (ThFabba) [reply sdk/lib/atl/atlbase.h]: But whatever the value of 'hr' is in this failure case wouldn't :p

**Pull request merged**